Sources/SwiftBuildSupport/SwiftBuildSystem.swift

```diff
@@ -174,6 +174,388 @@ private final class PlanningOperationDelegate: SWBPlanningOperationDelegate, Sen
     }
 }
 
+/// Convenience extensions to extract taskID and targetID from the LocationContext.
+extension SwiftBuildMessage.LocationContext {
+    var taskID: Int? {
+        switch self {
+        case .task(let id, _), .globalTask(let id):
+            return id
+        case .target, .global:
+            return nil
+        }
+    }
+
+    var targetID: Int? {
+        switch self {
+        case .task(_, let id), .target(let id):
+            return id
+        case .global, .globalTask:
+            return nil
+        }
+    }
+}
+
+/// Handler for SwiftBuildMessage events sent by the SWBBuildOperation.
+final class SwiftBuildSystemMessageHandler {
+    private let observabilityScope: ObservabilityScope
+    private let outputStream: OutputByteStream
+    private let logLevel: Basics.Diagnostic.Severity
+    private var buildState: BuildState = .init()
+    private var tasksEmitted: Set<String> = []
+
+    let progressAnimation: ProgressAnimationProtocol
+    var serializedDiagnosticPathsByTargetName: [String: [Basics.AbsolutePath]] = [:]
+
+    private var unprocessedDiagnostics: [SwiftBuildMessage.DiagnosticInfo] = []
+
+    public init(
+        observabilityScope: ObservabilityScope,
+        outputStream: OutputByteStream,
+        logLevel: Basics.Diagnostic.Severity
+    )
+    {
+        self.observabilityScope = observabilityScope
+        self.outputStream = outputStream
+        self.logLevel = logLevel
+        self.progressAnimation = ProgressAnimation.ninja(
+            stream: self.outputStream,
+            verbose: self.logLevel.isVerbose
+        )
+    }
+
+    struct BuildState {
+        private var targetsByID: [Int: SwiftBuild.SwiftBuildMessage.TargetStartedInfo] = [:]
+        private var activeTasks: [Int: SwiftBuild.SwiftBuildMessage.TaskStartedInfo] = [:]
+        private var taskBuffer: [String: Data] = [:]
```

Comment:
Instead of a string as the key, can you use a `SwiftBuildMessage.LocationContext`?